src/message_bus.zig

```diff
@@ -624,57 +629,81 @@ const Connection = struct {
-        }
-    }
-
-    fn recv_header(self: *Connection) void {
-        self.recv(
-            on_recv_header,
-            self.incoming_message.?.buffer[self.recv_progress..@sizeOf(Header)],
-        );
-    }
-
-    fn recv_body(self: *Connection) void {
-        self.recv(
-            on_recv_body,
-            self.incoming_message.?.buffer[self.recv_progress..self.incoming_message.?.header.size],
-        );
+            return;
+        };
+        self.recv();
     }
 
-    fn recv(
-        self: *Connection,
-        comptime callback: fn (*Connection, *IO.Completion, IO.RecvError!usize) void,
-        buffer: []u8,
-    ) void {
+    fn recv(self: *Connection) void {
         assert(self.peer != .none);
         assert(self.state == .connected);
         assert(self.fd != -1);
+
+        if (self.pipeline_message) |pipeline| {
+            assert(self.recv_progress != 0);
+
+            const buffer = pipeline.buffer[self.pipeline_start..self.recv_progress];
+            if (buffer.len >= @sizeOf(Header)) {
+                const header = mem.bytesAsValue(Header, buffer[0..@sizeOf(Header)]);
+                if (!header.valid_checksum()) {
+                    log.err("invalid checksum on header received from {}", .{self.peer});
+                    self.shutdown();
+                    return;
+                }
+
+                if (buffer.len >= header.size) {
+                    const body = buffer[@sizeOf(Header)..header.size];
+                    if (!header.valid_checksum_body(body)) {
+                        log.err("invalid checksum on body received from {}", .{self.peer});
+                        self.shutdown();
+                        return;
+                    }
+                    mem.copy(u8, self.incoming_message.?.buffer, buffer[0..header.size]);
+
+                    // TODO: rewrite ConcurrentRanges to not use async. This nosuspend is only
+                    // safe because we do not do any asynchronous disk IO yet.
+                    var frame = async self.message_bus.server.on_message(self.incoming_message.?);
+                    nosuspend await frame;
+
+                    self.message_bus.unref(self.incoming_message.?);
+                    self.incoming_message = null;
+
+                    if (header.size == self.recv_progress) {
+                        self.recv_progress = 0;
+                        self.message_bus.unref(pipeline);
+                        self.pipeline_message = null;
+                    }
+                    self.maybe_recv();
+                    return;
+                }
+            }
+            // The pipeline message is not yet fully received, copy to the new buffer and start
+            // a new recv operation.
+            mem.copy(u8, self.incoming_message.?.buffer, buffer);
+            self.recv_progress = buffer.len;
+            self.message_bus.unref(pipeline);
+            self.pipeline_message = null;
+        }
```

Comment:
What do you think of refactoring this duplicate header validation and `on_message` logic to be within a single `parse()` code path?

Then, over here we do a one-liner copy of `pipeline_message` into `recv_message`, like: https://github.com/coilhq/tigerbeetle/blob/master/src/main.zig#L183-L191

And then we do a branch to decide whether to `parse()` or `recv()`, like: https://github.com/coilhq/tigerbeetle/blob/master/src/main.zig#L243-L247

Or did I miss something?

Reply:
If I read that code correctly, messages are potentially copied more than once if multiple are received in a single recv operation as the `parse()` function only deals with a single message at a time. The current implementation on this branch will copy each message at most once, if at all.

I do agree that the duplicate header validation code paths are not ideal though, I'll see if there's a different way to refactor that away.

Reply:
Ah, yes, it's better to copy at most once. I had missed that.

It looks like there may be a bug here where we don't advance `self.pipeline_offset` always for each successive pipeline message?

We should also then rather use a loop rather than recursion to get another message rather than go back to `maybe_recv()` after each pipeline message.

And also refactor so that we only call `checksum_valid()` and `on_message()` from one call site.